virtcontainers/qemu_arm64.go

```diff
@@ -38,15 +41,41 @@ var kernelRootParams = []Param{
 	{"root", "/dev/vda1"},
 }
 
+// variables rather than consts to allow tests to modify them
+var interruptFile = "/proc/interrupts"
+
+//on aarch64, we support different gic interrupt controllers
+//maximum number of vCPUs depends on the GIC version, or on how
+//many redistributors we can fit into the memory map.
+var gicList = map[string]uint32{
+	"GICv2": uint32(8),
+	"GICv3": uint32(123),
+	"GICv4": uint32(123),
+}
+
 var supportedQemuMachines = []govmmQemu.Machine{
 	{
 		Type:    QemuVirt,
 		Options: defaultQemuMachineOptions,
 	},
 }
 
+// Logger returns a logrus logger appropriate for logging qemu-aarch64 messages
+func qemuArmLogger() *logrus.Entry {
+	return virtLog.WithField("subsystem", "qemu-aarch64")
+}
+
 // MaxQemuVCPUs returns the maximum number of vCPUs supported
 func MaxQemuVCPUs() uint32 {
+	bytes, err := ioutil.ReadFile(interruptFile)
+	if err != nil {
+		qemuArmLogger().WithField("interruptFile", interruptFile).WithError(err).Error("Failed to read interrupt file")
+	}
+	for gicType, vCPUs := range gicList {
+		if strings.Contains(string(bytes), gicType) {
+			return vCPUs
```

Comment:
I tested this PR in a system with GICv3 and 96 cores. This function returns `123`, unfortunately it won't be honoured since the maximum number of vCPUs can't be exceeded https://github.com/kata-containers/runtime/blob/14bcd697b717d078f285439810070891fb6a1c7f/cli/config.go#L215-L232
My concern with this patch is that the actual number of physical cores will be exceeded and the memory footprint will be big (again) see https://github.com/kata-containers/runtime/commit/07db945b0976cac912c78d57efe563672267c1f2

Reply:
@Pennyzct have you tested this change in a system with GIC >= 3 and physical CPUs > 123 ?

Reply:
Hi~ @devimc  That's how I found the problem and pulled this pr😊. I was installing and running kata in new ThunderX II which contains 224 physical cpu cores.
```
:~# lscpu
Architecture:          aarch64
Byte Order:            Little Endian
CPU(s):                224
On-line CPU(s) list:   0-223
Thread(s) per core:    4
Core(s) per socket:    28
Socket(s):             2
NUMA node(s):          2
NUMA node0 CPU(s):     0-111
NUMA node1 CPU(s):     112-223
```
Err occurred and outputs were as follows:
`docker: Error response from daemon: OCI runtime create failed: qemu-system-aarch64: Number of SMP CPUs requested (224) exceeds max CPUs supported by machine 'mach-virt' (123): unknown`
when applying my patch, vCPUs will be reduced into `123` in func `defaultMaxVCPUs` as you mentioned above and kata will run successfully.